system/Test/FilterTestTrait.php

```diff
@@ -0,0 +1,288 @@
+<?php
+
+/**
+ * This file is part of the CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Test;
+
+use CodeIgniter\Filters\Exceptions\FilterException;
+use CodeIgniter\Filters\Filters;
+use CodeIgniter\Filters\FilterInterface;
+use CodeIgniter\HTTP\RequestInterface;
+use CodeIgniter\HTTP\ResponseInterface;
+use CodeIgniter\Router\RouteCollection;
+use Config\Filters as FiltersConfig;
+use Config\Services;
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * Filter Test Trait
+ *
+ * Provides functionality for testing
+ * filters and their route associations.
+ *
+ * @mixin CIUnitTestCase
+ */
+trait FilterTestTrait
+{
+	/**
+	 * Have the one-time classes been instantiated?
+	 *
+	 * @var boolean
+	 */
+	private $doneFilterSetUp = false;
+
+	/**
+	 * The active IncomingRequest or CLIRequest
+	 *
+	 * @var RequestInterface
+	 */
+	protected $request;
+
+	/**
+	 * The active Response instance
+	 *
+	 * @var ResponseInterface
+	 */
+	protected $response;
+
+	/**
+	 * The Filters configuration to use.
+	 * Extracted for access to aliases
+	 * during Filters::discoverFilters().
+	 *
+	 * @var FiltersConfig|null
+	 */
+	protected $filtersConfig;
+
+	/**
+	 * The prepared Filters library.
+	 *
+	 * @var Filters|null
+	 */
+	protected $filters;
+
+	/**
+	 * The default App and discovered
+	 * routes to check for filters.
+	 *
+	 * @var RouteCollection|null
+	 */
+	protected $collection;
+
+	//--------------------------------------------------------------------
+	// Staging
+	//--------------------------------------------------------------------
+
+	/**
+	 * Initializes dependencies once.
+	 *
+	 * @return void
+	 */
+	protected function setUpFilterTestTrait(): void
+	{
+		if ($this->doneFilterSetUp === false)
```

Comment:
how about flip the if and `return; `early?

```php
if ($this->doneFilterSetUp === true)
{
    return;
}

// Create our own Request and Response so we can
```
by that, the deep if can be reduced.

Reply:
Yes good point. I used to have some stuff outside that `if` but now it controls the whole method. Will change.